### PR TITLE
Build Library and Binary sources next to the output

### PR DESCRIPTION
**Pull request by ptia**
Currently, the `Binary` rule compiles sources into .o files in the output directory corresponding to where the sources are found, and there is no way to change this. These .o files are artifacts that can only be used by the `Binary` rule that generates them, because the `ObjectFile` target that creates them is not visible anywhere else.
If the source files are not close to where the Binary target is defined, this means that artifacts that are only useful to the `Binary` target are built somewhere else in the output directory. This is a problem because:
 * It breaks locality: if I build something in a directory I expect all its artifacts to end up next to the final target
 * It means that if two different `Binary` targets want to build the same .cc files (for example two different cross-compilation targets), they'll try to generate the same .o files, and the build will fail

 The solution is to build all `ObjectFiles` for a Binary in a subdirectory next to its final target. This does not change the name of the .o files, or affect existing code in any way other than changing the internal layout of the BUILD directory.

This also applies to the `Library` rule.

### CODE REVIEW

**Comment by ptia** [comment]
Reason for this is to make libsupcxx cross-compilation without flags much easier

**Comment by ptia** [comment]
In other words: `Library` and `Binary`  create .o files that only they can see, potentially very far from the target that needs them, and with nothing to indicate that these files are artifacts of the `Library` and `Binary` targets. This makes cross-compilation very difficult

File: RULES/cc/cc.go [hunk 1]
```diff
@@ -33,7 +36,11 @@ func (obj ObjectFile) Build(ctx core.Context) {
 }
 
 func (obj ObjectFile) out() core.OutPath {
-	return obj.Src.WithExt("o")
+	defaultOut := obj.Src.WithExt("o")
+	if obj.OutDir == nil {
+		return defaultOut
+	}
+	return obj.OutDir.WithSuffix("/"+defaultOut.Relative())
```
Comment (jfrohnhofen) [inline RULES/cc/cc.go:43]: doesn't this duplicate the entire path? E.g.: `defaultOut` is `foo/bar/baz.o` and `obj.OutDir` is `foo/bar/mylib.a-OBJS`, won't we get `foo/bar/mylib.a-OBJS/foo/bar/baz.o`? Is that intentional? Is my understanding wrong?

Comment (ptia) [reply RULES/cc/cc.go:43]: Yes it does, it's to avoid files with the same object name in different paths from conflicting

Comment (ptia) [reply RULES/cc/cc.go:43]: It's very much an edge case, and two object files with the same name would probably also confuse something like a linker script, but I wanted to be robust.

Comment (jfrohnhofen) [reply RULES/cc/cc.go:43]: Make sense

**Review by jfrohnhofen — APPROVED** [review]

**Pull request merged**